3rdparty/libjpeg-turbo/CMakeLists.txt

```diff
@@ -15,8 +15,68 @@ endif()
 
 message(STATUS "libjpeg-turbo: VERSION = ${VERSION}, BUILD = ${BUILD}")
 
+# Detect CPU type and whether we're building 64-bit or 32-bit code
+math(EXPR BITS "${CMAKE_SIZEOF_VOID_P} * 8")
+string(TOLOWER ${CMAKE_SYSTEM_PROCESSOR} CMAKE_SYSTEM_PROCESSOR_LC)
+set(COUNT 1)
+foreach(ARCH ${CMAKE_OSX_ARCHITECTURES})
+  if(COUNT GREATER 1)
+    message(FATAL_ERROR "The libjpeg-turbo build system does not support multiple values in CMAKE_OSX_ARCHITECTURES.")
+  endif()
+  math(EXPR COUNT "${COUNT}+1")
+endforeach()
+if(CMAKE_SYSTEM_PROCESSOR_LC MATCHES "x86_64" OR
+        CMAKE_SYSTEM_PROCESSOR_LC MATCHES "amd64" OR
+        CMAKE_SYSTEM_PROCESSOR_LC MATCHES "i[0-9]86" OR
+        CMAKE_SYSTEM_PROCESSOR_LC MATCHES "x86" OR
+        CMAKE_SYSTEM_PROCESSOR_LC MATCHES "ia32")
+  if(BITS EQUAL 64 OR CMAKE_C_COMPILER_ABI MATCHES "ELF X32")
+    set(CPU_TYPE x86_64)
+  else()
+    set(CPU_TYPE i386)
+  endif()
+  if(NOT CMAKE_SYSTEM_PROCESSOR STREQUAL ${CPU_TYPE})
+    set(CMAKE_SYSTEM_PROCESSOR ${CPU_TYPE})
+  endif()
+elseif(CMAKE_SYSTEM_PROCESSOR_LC STREQUAL "aarch64" OR
+        CMAKE_SYSTEM_PROCESSOR_LC MATCHES "^arm")
+  if(BITS EQUAL 64)
+    set(CPU_TYPE arm64)
+  else()
+    set(CPU_TYPE arm)
+  endif()
+elseif(CMAKE_SYSTEM_PROCESSOR_LC MATCHES "^ppc" OR
+        CMAKE_SYSTEM_PROCESSOR_LC MATCHES "^powerpc")
+  set(CPU_TYPE powerpc)
+else()
+  set(CPU_TYPE ${CMAKE_SYSTEM_PROCESSOR_LC})
+endif()
+if(CMAKE_OSX_ARCHITECTURES MATCHES "x86_64" OR
+        CMAKE_OSX_ARCHITECTURES MATCHES "arm64" OR
+        CMAKE_OSX_ARCHITECTURES MATCHES "i386")
+  set(CPU_TYPE ${CMAKE_OSX_ARCHITECTURES})
+endif()
+if(CMAKE_OSX_ARCHITECTURES MATCHES "ppc")
+  set(CPU_TYPE powerpc)
+endif()
+if(MSVC_IDE AND CMAKE_GENERATOR_PLATFORM MATCHES "arm64")
+  set(CPU_TYPE arm64)
+endif()
```

Comment:
OpenCV already has processor options detector in CMake. Please reuse variables like X86_64, ARM, PPC64 and others from https://github.com/opencv/opencv/blob/4.x/cmake/OpenCVDetectCXXCompiler.cmake.

Reply:
Pay attention that NEON, AVX512, etc support are managed with dedicated flags.

Reply:
> Pay attention that NEON, AVX512, etc support are managed with dedicated flags.

Could you eloborate more this? I did not get it what you mean by that.

Reply:
https://github.com/opencv/opencv/wiki/CPU-optimizations-build-options

Reply:
Briefly, NEON,  AVX2, AVX512, etc support are not guaranteed on all CPUs. There are Atom chips, even without AVX2 support. The same for NEON. Some NVIDIA Tegra chips has arm-v7a architecture, but does not implement NEON extension. NEON code will raise SIGILL here.
`CPU_BASELINE_FINAL` contains all hardware specific optimizations that are enabled by default and does not participate in dynamic dispatch. libjpeg turbo should use the list of options to enable proper optimizations.

Reply:
For example you can take a look on Carotene HAL management for ARM: https://github.com/opencv/opencv/blob/9fcf0152140c449f74678add820e81de3e69b8b0/CMakeLists.txt#L880

Reply:
If I understand your concern correctly, NEON and others are handled by libjpeg-turbo simd/cmakelists. 

For example, [here](https://github.com/opencv/opencv/blob/79afd80817c8d84484052ae06e64d61db202d27a/3rdparty/libjpeg-turbo/src/simd/CMakeLists.txt#L218) NEON support are handled below. 

I will investigave further. Thanks for the link.

Reply:
If this code is borrowed from upstream, then we should keep it with minimal changes (to keep updates from upstream reliable)

Reply:
I borrowed from upstream. It makes sense to keep it as original. What do you mean by minimal changes?